C51-62: Integrate functionality next milestone card and next activity card

swastikpareek opened the pull request:
## Overview 
This PR contains basic UI enhancements and integration of next activity and next milestone card

## Before 

<img width="1680" alt="c51-62 - before" src="https://user-images.githubusercontent.com/3340537/45827518-35898280-bd14-11e8-980b-a832823fc7b8.png">


## After
<img width="1680" alt="c51-62 - loading state" src="https://user-images.githubusercontent.com/3340537/45827429-0a069800-bd14-11e8-98bc-b81671bf08ca.png">
<img width="1057" alt="c51-62 - no result" src="https://user-images.githubusercontent.com/3340537/45827430-0a069800-bd14-11e8-9fc0-4f6491cc71c9.png">

<img width="1017" alt="c51-62- after the most recent" src="https://user-images.githubusercontent.com/3340537/45880211-d46db780-bdc4-11e8-97b8-324dfdb4a720.png">


### Activity card states
 * With icon and type = task 
  <img width="304" alt="c51-62 - act with icon and task" src="https://user-images.githubusercontent.com/3340537/45870331-408df280-bda8-11e8-84a2-b35fa68ebf47.png">

 * With icon 
   <img width="314" alt="c51-62 - activity with icon" src="https://user-images.githubusercontent.com/3340537/45870332-41268900-bda8-11e8-90d9-21d2ef2c090c.png">

 * With task
  <img width="304" alt="c51-62 - activity with task" src="https://user-images.githubusercontent.com/3340537/45870333-41268900-bda8-11e8-8f60-d12fffe850b9.png">

 * Without icon and task
  <img width="314" alt="c51-62 - activity without icon" src="https://user-images.githubusercontent.com/3340537/45870334-41bf1f80-bda8-11e8-92bf-870f8454fe0a.png">

## Technical Details

* Moved `isOverdue` and `formatDate` function into a new `DateHelper` service and added unit tests for them.
* Added three different variants for activity card 
  *  Activity Card Big - `ActivityCard--Big.html` and `_activity-card--big.scss`
  *  Activity Card Long - `ActivityCard--Long.html` and `_activity-card--long.scss`
  *  Activity Card Short - `ActivityCard--Short.html`  and `_activity-card--short.scss`
* Added two new attributes to activity card directive
 * `mode` - To switch between above mentioned visual modes - default to 'short'
 * `type` - To set the inner styles and no result behaviours of different types of activity the cards.

* Added new classes
  * `civicase__activity-card--big--milestone` : For milestone specific css


* Added empty state for actitivity card styles inside `_activity-no-result.scss` and new classes added
  * `civicase__no-result-icon` `civicase__no-result-icon--milestone` `civicase__activity-no-result-icon--activity`, `civicase__activity-no-result-icon--communication`, `civicase__activity-no-result-icon--tasks`.

* Update activity return params to fetch more required activity fields.

* Added tags related info in `GetDetails.php` file to get tags related info in activity cards.

* Changed `civicase__big-date` class convention to `civicase__highlighted-date` class
* Added hover styling to cards
* Fix case view tabs panel z-index and glicth issue.
* New classe `civicase__activity-card-left` has been added for taking care of left section of big activity card.

## Updates
 * Fix checkbox check issue for next activity being a task 
 * Add `star`  and `attachment` icons to the next activity and next mileston 
 * Theme the dropdown with the icons and styles
 * Update logic to each activity card which is milestone to have a red ribbon as a background .
<img width="1246" alt="c51-62 - after setting the label for milestone" src="https://user-images.githubusercontent.com/3340537/45880191-c61f9b80-bdc4-11e8-9b32-a646dcb965f9.png">

deb1990 commented:
UI Issues
1. On hover, there is a blue border, is it in the design?
![2018-09-21 at 11 09 am](https://user-images.githubusercontent.com/5058867/45862305-c9983000-bd8e-11e8-948c-b220c6090fe6.png)
we should follow the above for hover, unless specified

2. 3 dots design does not match with mock. Maybe we are using wrong icon and need to use from material.

3. When next activity is of task type, do we get the checkbox? please confirm

4. When there is no icon set for the next activity, how does it look?

5. Have we heard anything from Jon about the "Star" and "Attachment"?

deb1990 commented on ang/civicase/Utils.js:
```diff
@@ -623,6 +623,31 @@
+     * @param {String} format ISO string
+     * @return {String} the formatted date
+     */
+    this.formatDate = function (date, format = 'DD/MM/YYYY') {
```
`format = 'DD/MM/YYYY'` This is ES6 , which we are not using for now. Please use ES5.
```
this.formatDate = function (date, format) {
      format = format || 'DD/MM/YYYY';
      
      return moment(date).format(format);
    };
```

deb1990 commented on ang/test/civicase/Utils.spec.js:
```diff
@@ -0,0 +1,35 @@
+/* eslint-env jasmine */
```
According to the above comment, the file name should change.

deb1990 commented on ang/test/civicase/Utils.spec.js:
```diff
@@ -0,0 +1,35 @@
+      }));
+
+      describe('formatDate()', function () {
+        it('checks the format of the date', function () {
```
Wrong description.
It does not check the format, it returns the date in the sent format.

deb1990 commented on ang/test/civicase/Utils.spec.js:
```diff
@@ -0,0 +1,35 @@
+          futureDate = moment().add(7, 'd');
+        });
+
+        it('checks the date is overdue ', function () {
```
`returns true when date is in past`

deb1990 commented on ang/test/civicase/Utils.spec.js:
```diff
@@ -0,0 +1,35 @@
+          expect(DateHelper.isOverdue(pastDate)).toBe(true);
+        });
+
+        it('checks the date is not overdue ', function () {
```
`returns false when date is not in past`

deb1990 commented on icons/activities.svg:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Now that we have different kinds of resources. Its better to collect them under a folder.
So please create a folder `resource`. And move `fonts` and `icons` inside.

deb1990 commented on scss/components/_big-date.scss:
```diff
@@ -1,5 +1,5 @@
 /** Big calendar type date styling **/
-.civicase__big-date {
+.civicase__highlighted-date {
```
File name and comments should change to follow this changed component name.

deb1990 commented on scss/components/_activity-no-result-icon.scss:
```diff
@@ -0,0 +1,15 @@
+.civicase__no-result-icon {
```
These icons are only for activity right?
The component name does not denote that.
Should be 
`civicase__activity-no-result-icon` - this one should have the SVG for Activities which is not milestone.
`civicase__activity-no-result-icon--milestone`
`civicase__activity-no-result-icon--task` (no need to create now, but for future)
`civicase__activity-no-result-icon--communication` (no need to create now, but for future)

Lastly, this is not a new component, please place them under `ActivityCard.scss`

swastikpareek replied:
> Should be `civicase__activity-no-result-icon` - Noted 
>  this is not a new component, please place them under `ActivityCard.scss` - Noted

> this one should have the SVG for Activities which is not milestone. 
I think according to BEM. There should be a base class which contains all the basic generic styles
and then these styles should have different variants.
so it should be something like this
```
.civicase__activity-no-result-icon {
  background-position: center center;
  background-repeat: no-repeat;
  background-size: contain;
  height: 48px;
  width: 48px;
}

.civicase__activity-no-result-icon--milestone {
  background-image: url('../resources/icons/milestone.svg');
}

.civicase__activity-no-result-icon--activity {
  background-image: url('../resources/icons/activities.svg');
}

.civicase__activity-no-result-icon--activity {
  background-image: url('../resources/icons/activities.svg');
}

.civicase__activity-no-result-icon--activity {
  background-image: url('../resources/icons/activities.svg');
}

```

deb1990 commented on scss/components/_activity-card.scss:
```diff
@@ -111,9 +110,6 @@
 .civicase__activity-subject {
   color: $gray-darker;
   font-weight: $font-weight-heading;
-  overflow: hidden;
```
Why are these removed?

swastikpareek replied:
These are not removed. These are moved under `_activity-card--short.scss` as we need these styles for short activity card variant only.

swastikpareek commented on ang/civicase/CaseDetails.js:
```diff
@@ -334,9 +334,34 @@
+      _.each(activities, function (act, key) {
+        if (CRM.civicase.activityTypes[act.activity_type_id].grouping !== 'milestone' && !hit) {
+          nextActivity = act;
+          nextActivity.type = CRM.civicase.activityTypes[act.activity_type_id].label;
```
`Next Milestone` and `next_activity` are pre-populated with these keys from `case/GetDetails.php` file. If we wish to do it for `activities` coming from another Backend API route. We will need to do it at the backend and will require more efforts and time. And also as per our discussion (which we follow for civiHR development) its better to manipulate data at the front end.

Please let me know if you have any doubts.

swastikpareek commented:
@deb1990 
 
> 1. On hover, there is a blue border, is it in the design?
>    ![2018-09-21 at 11 09 am](https://user-images.githubusercontent.com/5058867/45862305-c9983000-bd8e-11e8-948c-b220c6090fe6.png)
>    we should follow the above for hover, unless specified
> 2. 3 dots design does not match with mock. Maybe we are using wrong icon and need to use from material.
> 3. When next activity is of task type, do we get the checkbox? please confirm
> 4. When there is no icon set for the next activity, how does it look?
> 5. Have we heard anything from Jon about the "Star" and "Attachment"?

1. Hover effect is updated
2. Dot icon now being used from material design. Please refer screenshot.
3. Please find the attached screenshots for it.
4. Same as 3.
5. Yes. Jon has over-sighted this he said. It is good to have these on the card but we need to get this clarified from @shitijg - he added.
 As per our discussion we can have this discussed later w/o blocking this PR.

deb1990 commented on ang/civicase/ActivityCard.js:
```diff
@@ -4,21 +4,33 @@
+      templateUrl: function (elem, attrs) {
+        switch (attrs.mode) {
+          case 'big':
+            return `~/civicase/ActivityCard--Big.html`;
```
You have used ` instead of `'`

deb1990 commented on ang/civicase/ActivityCard.js:
```diff
@@ -4,21 +4,33 @@
     $scope.activityFeedUrl = getActivityFeedUrl;
     $scope.formatDate = CRM.utils.formatDate;
     $scope.templateExists = templateExists;
+    $scope.formatDate = DateHelper.formatDate;
```
`$scope.formatDate` function is defined twice. Also on line 30

deb1990 commented on ang/civicase/CaseListTable.js:
```diff
@@ -241,30 +241,34 @@
+          'is_overdue', 'case_id'
+        ]
+      };
       var returnParams = {
```
This should be renamed to `returnCaseParams` to avoid confusion.

deb1990 commented on ang/civicase/DateHelperService.js:
```diff
@@ -0,0 +1,30 @@
+    };
+
+    /**
+     * Formats Date in correct format (DD/MM/YYYY)
```
```
* Formats Date in sent format
* Default format is (DD/MM/YYYY)
```

deb1990 commented on ang/test/civicase/DateHelperService.spec.js:
```diff
@@ -0,0 +1,35 @@
+/* eslint-env jasmine */
+(function ($) {
+  describe('Utils', function () {
```
Utils should be removed as its not part of Utils anymore.

deb1990 commented on scss/components/_activity-card.scss:
```diff
@@ -216,3 +216,27 @@
+}
+
+.civicase__activity-no-result-icon--communication {
+  background-image: url('../resources/icons/activities.svg');
```
Wrong svg?

deb1990 commented on scss/components/_activity-card.scss:
```diff
@@ -216,3 +216,27 @@
+}
+
+.civicase__activity-no-result-icon--tasks {
+  background-image: url('../resources/icons/activities.svg');
```
Wrong svg?

deb1990 commented:
@swastikpareek When you Delete Next Activity or Milestone, does the UI refresh to show latest data?

deb1990 commented:
@swastikpareek There is no hover effect on the 3 dot icons. I was there previously.

swastikpareek commented:
> @swastikpareek There is no hover effect on the 3 dot icons. I was there previously. 

Updated the styles. Please test

> @swastikpareek When you Delete Next Activity or Milestone, does the UI refresh to show latest data?

Fixed this as well
